File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-rec-iter-tail.rb
# LeetCode #94 - Binary Tree Inorder Traversal
# https://leetcode.com/problems/binary-tree-inorder-traversal/
# Iteratively implemented recursive solution with tail-call elimination.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
  out = []
  stack = []
  stack << Frame.new(root, :go_left) if root

  until stack.empty?
    frame = stack.last

    case frame.state
    when :go_left
      stack << Frame.new(frame.node.left, :go_left) if frame.node.left
      frame.state = :go_right
    when :go_right
      out << frame.node.val
      stack.pop
      stack << Frame.new(frame.node.right, :go_left) if frame.node.right
    end
  end

  out
end

# A stack frame for iteratively implemented recursive binary tree traversal.
class Frame
  attr_reader :node
  attr_accessor :state

  def initialize(node, state)
    @node = node
    @state = state
  end
end

File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-rec-iter-tail.java
// LeetCode #94 - Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/
// Iterative implementation of right-side tail recursive algorithm
// (state machine, 2 states instead of 3).

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        inorder(root, values::add);
        return values;
    }

    private static void noop(int _value) {
        // Do nothing.
    }

    private static enum State {
        GO_LEFT,
        GO_RIGHT
    }

    private static final class Frame {
        Frame(TreeNode node) {
            this.node = node;
            this.state = State.GO_LEFT;
        }

        final TreeNode node;
        State state;
    }

    private static void inorder(TreeNode root, IntConsumer action) {
        if (root == null) return;

        var stack = Collections.asLifoQueue(new ArrayDeque<Frame>());
        stack.add(new Frame(root));

        while (!stack.isEmpty()) {
            var frame = stack.element();

            switch (frame.state) {
            case GO_LEFT:
                frame.state = State.GO_RIGHT;
                if (frame.node.left != null)
                    stack.add(new Frame(frame.node.left));
                break;

            case GO_RIGHT:
                action.accept(frame.node.val);
                stack.remove();
                if (frame.node.right != null)
                    stack.add(new Frame(frame.node.right));
                break;
            }
        }
    }
}
